File: typescript/src/checks/effects_bypass.ts
/**
 * The built-in `effects-bypass` check provider.
 *
 * It statically analyses the consumer's own sources and fails on any direct
 * process, filesystem-mutation or network call reachable from a handler that
 * opted into `ctx.effects`. Opting in is the trigger: a function that uses the
 * effects handle must route ALL of its effects through it, or the preview it
 * promises is a lie.
 *
 * Additionally -- and this part is TypeScript-specific -- it flags the two
 * accepted Proxy ceilings: a bare truthiness test and an identity comparison
 * against a value the analyser can trace to an effects-handle return. Those two
 * are the only things the runtime seal cannot catch, so lint is the sole line
 * of defence and the check names them explicitly.
 *
 * ANALYSER: the TypeScript compiler API, a REGULAR dependency -- `typescript`
 * sits in `dependencies`, not `devDependencies`, and there is no optional
 * import and no soft degradation. Concretely it is the compiler's own scanner
 * (`createScanner` from `typescript/unstable/ast`). The compiler package at
 * version 7 is the native port: it ships the scanner, the SyntaxKind enum and
 * the AST node predicates, but NO in-process parser -- building a syntax tree
 * there means spawning the native language server against a resolved tsconfig,
 * which a check declared `fast` and `pure` must not do. The scope rule below is
 * therefore expressed over brace blocks rather than function nodes: a call is a
 * finding when some brace block contains BOTH the call and an `.effects`
 * mention, which is the token-level reading of "reachable from a handler that
 * opted in".
 */

import { readdirSync, readFileSync, statSync } from "node:fs";
import { join, relative } from "node:path";
import {
	computeLineStarts,
	createScanner,
	SyntaxKind,
} from "typescript/unstable/ast";
import type { AppImpl } from "../app.js";
import { type CheckSpec, errorCheckSpec } from "./provider.js";

/** Process starts. Matched on the called name, bare or through a receiver. */
const BYPASS_PROCESS: ReadonlySet<string> = new Set([
	"spawn",
	"spawnSync",
	"exec",
	"execSync",
	"execFile",
	"execFileSync",
	"fork",
]);

/** Filesystem mutations (sync and promise forms). */
const BYPASS_FILESYSTEM: ReadonlySet<string> = new Set([
	"writeFileSync",
	"appendFileSync",
	"mkdirSync",
	"mkdtempSync",
	"rmSync",
	"rmdirSync",
	"unlinkSync",
	"renameSync",
	"chmodSync",
	"chownSync",
	"symlinkSync",
	"linkSync",
	"truncateSync",
	"copyFileSync",
	"cpSync",
	"createWriteStream",
	"writeFile",
	"appendFile",
	"mkdir",
	"mkdtemp",
	"rm",
	"rmdir",
	"unlink",
	"rename",
	"chmod",
	"chown",
	"symlink",
	"link",
	"truncate",
	"copyFile",
	"cp",
]);

/** Network calls, banned only through the receivers below. */
const BYPASS_NETWORK: ReadonlySet<string> = new Set([
	"request",
	"get",
	"post",
	"put",
	"patch",
	"delete",
	"head",
	"createConnection",
	"connect",
]);

/**
 * Network members are banned only through these receivers, so an ordinary
 * `map.get(...)` inside a handler is not a finding.
 */
const BYPASS_NETWORK_RECEIVERS: ReadonlySet<string> = new Set([
	"http",
	"https",
	"axios",
	"got",
	"undici",
	"net",
	"tls",
	"request",
	"client",
	"session",
	"agent",
]);

const SKIP_DIRS: ReadonlySet<string> = new Set([
	"node_modules",
	"dist",
	"dist-test",
	"build",
	"out",
	"coverage",
	"vendor",
]);

const SOURCE_EXTS = [".ts", ".tsx", ".mts", ".cts", ".js", ".mjs", ".cjs"];

/** One reported bypass. */
export interface BypassFinding {
	readonly file: string;
	readonly line: number;
	readonly text: string;
	readonly kind: "call" | "ceiling";
}

/** Collects source files under `root`, in directory-then-name order. */
function collectSourceFiles(root: string): string[] {
	const out: string[] = [];
	const walk = (dir: string): void => {
		let entries: string[];
		try {
			entries = readdirSync(dir).sort();
		} catch {
			return;
		}
		const files: string[] = [];
		const dirs: string[] = [];
		for (const entry of entries) {
			if (entry.startsWith(".")) {
				continue;
			}
			const full = join(dir, entry);
			let isDir = false;
			try {
				isDir = statSync(full).isDirectory();
			} catch {
				continue;
			}
			if (isDir) {
				if (!SKIP_DIRS.has(entry)) {
					dirs.push(full);
				}
			} else if (SOURCE_EXTS.some((ext) => entry.endsWith(ext))) {
				files.push(full);
			}
		}
		out.push(...files);
		for (const d of dirs) {
			walk(d);
		}
	};
	walk(root);
	return out;
}

interface Tok {
	readonly kind: SyntaxKind;
	readonly text: string;
	readonly start: number;
}

/**
 * Tokenizes with the compiler's scanner, dropping trivia.
 *
 * The scanner is a pure lexer: it cannot leave a template-substitution state on
 * its own, because in a real compile the PARSER decides when a `}` closes a
 * `${` and calls back into `reScanTemplateToken`. Without that cooperation the
 * scanner stalls at the closing brace, returning a zero-width token forever.
 * Progress is therefore asserted explicitly: on a stall, re-scan as a template
 * continuation, and if even that does not advance, step one character and carry
 * on. Both recoveries are lossless for this analyser, which reads names,
 * punctuation and brace depth -- never template contents.
 */
function tokenize(text: string): Tok[] {
	const scanner = createScanner(/* skipTrivia */ true);
	scanner.setText(text);
	const toks: Tok[] = [];
	let lastEnd = -1;
	for (;;) {
		let kind = scanner.scan();
		if (kind === SyntaxKind.EndOfFile) {
			break;
		}
		if (scanner.getTokenEnd() <= lastEnd) {
			kind = scanner.reScanTemplateToken(/* isTaggedTemplate */ false);
			if (scanner.getTokenEnd() <= lastEnd) {
				if (lastEnd + 1 >= text.length) {
					break;
				}
				scanner.resetTokenState(lastEnd + 1);
				lastEnd += 1;
				continue;
			}
		}
		lastEnd = scanner.getTokenEnd();
		toks.push({
			kind,
			text: scanner.getTokenText(),
			start: scanner.getTokenStart(),
		});
	}
	return toks;
}

function lineOf(lineStarts: readonly number[], pos: number): number {
	let lo = 0;
	let hi = lineStarts.length - 1;
	while (lo < hi) {
		const mid = (lo + hi + 1) >> 1;
		if ((lineStarts[mid] as number) <= pos) {
			lo = mid;
		} else {
			hi = mid - 1;
		}
	}
	return lo + 1;
}

/** True when the token is an identifier or a keyword usable as a member name. */
function isNameToken(t: Tok | undefined): boolean {
	return (
		t !== undefined &&
		/^[A-Za-z_$][A-Za-z0-9_$]*$/.test(t.text) &&
		t.kind !== SyntaxKind.StringLiteral
	);
}

/**
 * Scans one tokenized file. A call is a finding when some brace block contains
 * both the call site and an `.effects` mention.
 */
export function scanTokens(
	toks: readonly Tok[],
	text: string,
	rel: string,
): BypassFinding[] {
	const lineStarts = computeLineStarts(text);
	const findings: BypassFinding[] = [];

	// Brace blocks: blockOf[i] is the index of the innermost block containing
	// token i; blocks are (open, close) index pairs.
	const openStack: number[] = [];
	const blockParent: number[] = []; // per token: enclosing block id, or -1
	const blockOpenIdx: number[] = [];
	const blockEnclosing: number[] = [];
	const blockOfToken: number[] = new Array(toks.length).fill(-1);
	for (let i = 0; i < toks.length; i++) {
		const t = toks[i] as Tok;
		if (t.kind === SyntaxKind.CloseBraceToken) {
			openStack.pop();
		}
		const current =
			openStack.length > 0 ? (openStack[openStack.length - 1] as number) : -1;
		blockOfToken[i] = current;
		if (t.kind === SyntaxKind.OpenBraceToken) {
			const id = blockOpenIdx.length;
			blockOpenIdx.push(i);
			blockEnclosing.push(current);
			blockParent.push(current);
			openStack.push(id);
		}
	}

	// Blocks that (transitively) contain an `.effects` mention: for each
	// occurrence, every enclosing block opts in -- the token-level reading of
	// "every enclosing function opts in".
	const optedIn = new Set<number>();
	for (let i = 1; i < toks.length; i++) {
		if (
			(toks[i] as Tok).text === "effects" &&
			(toks[i - 1] as Tok).kind === SyntaxKind.DotToken
		) {
			let b = blockOfToken[i] as number;
			while (b !== -1) {
				optedIn.add(b);
				b = blockEnclosing[b] as number;
			}
		}
	}
	const inOptedInScope = (i: number): boolean =>
		optedIn.has(blockOfToken[i] as number);
	if (optedIn.size === 0) {
		return findings;
	}

	// Identifiers bound to an effects-handle return, so the Proxy-ceiling
	// checks have something concrete to trace.
	const carriers = new Set<string>();
	for (let i = 0; i + 2 < toks.length; i++) {
		const t = toks[i] as Tok;
		if (
			t.kind !== SyntaxKind.ConstKeyword &&
			t.kind !== SyntaxKind.LetKeyword &&
			t.kind !== SyntaxKind.VarKeyword
		) {
			continue;
		}
		const nameTok = toks[i + 1] as Tok;
		if (!isNameToken(nameTok)) {
			continue;
		}
		// Walk the initializer to the end of the statement looking for
		// `.effects.<method>(`.
		let j = i + 2;
		let sawEquals = false;
		let sawEffectsCall = false;
		for (; j < toks.length; j++) {
			const tj = toks[j] as Tok;
			if (tj.kind === SyntaxKind.SemicolonToken) {
				break;
			}
			if (tj.kind === SyntaxKind.EqualsToken) {
				sawEquals = true;
				continue;
			}
			if (
				sawEquals &&
				tj.text === "effects" &&
				(toks[j - 1] as Tok).kind === SyntaxKind.DotToken &&
				(toks[j + 1] as Tok | undefined)?.kind === SyntaxKind.DotToken &&
				(toks[j + 3] as Tok | undefined)?.kind === SyntaxKind.OpenParenToken
			) {
				sawEffectsCall = true;
			}
		}
		if (sawEquals && sawEffectsCall) {
			carriers.add(nameTok.text);
		}
	}

	const isCarrierAt = (i: number): boolean =>
		i >= 0 && i < toks.length && carriers.has((toks[i] as Tok).text);

	const push = (i: number, kind: "call" | "ceiling", text: string): void => {
		findings.push({
			file: rel,
			line: lineOf(lineStarts, (toks[i] as Tok).start),
			kind,
			text,
		});
	};

	for (let i = 0; i < toks.length; i++) {
		const t = toks[i] as Tok;
		if (!inOptedInScope(i)) {
			continue;
		}

		// --- direct effect calls ---
		if (
			isNameToken(t) &&
			(toks[i + 1] as Tok | undefined)?.kind === SyntaxKind.OpenParenToken
		) {
			const prevDot =
				(toks[i - 1] as Tok | undefined)?.kind === SyntaxKind.DotToken;
			const receiver = prevDot
				? (toks[i - 2] as Tok | undefined)?.text
				: undefined;
			// Anything reached through `.effects.` is exactly what we want.
			const throughEffects = prevDot && receiver === "effects";
			if (!throughEffects) {
				const leaf = t.text;
				const target = receiver !== undefined ? `${receiver}.${leaf}` : leaf;
				const banned =
					BYPASS_PROCESS.has(leaf) ||
					BYPASS_FILESYSTEM.has(leaf) ||
					(leaf === "fetch" && !prevDot) ||
					(BYPASS_NETWORK.has(leaf) &&
						receiver !== undefined &&
						BYPASS_NETWORK_RECEIVERS.has(receiver));
				if (banned) {
					push(
						i,
						"call",
						`calls ${target} directly; route it through ctx.effects`,
					);
					continue;
				}
			}
		}

		// --- Proxy ceiling 1: truthiness ---
		const truthyHere =
			// if (C) / while (C)
			((t.kind === SyntaxKind.IfKeyword ||
				t.kind === SyntaxKind.WhileKeyword) &&
				(toks[i + 1] as Tok | undefined)?.kind === SyntaxKind.OpenParenToken &&
				isCarrierAt(i + 2) &&
				(toks[i + 3] as Tok | undefined)?.kind ===
					SyntaxKind.CloseParenToken) ||
			// !C
			(t.kind === SyntaxKind.ExclamationToken && isCarrierAt(i + 1)) ||
			// C ? / C && / C ||
			(isCarrierAt(i) &&
				[
					SyntaxKind.QuestionToken,
					SyntaxKind.AmpersandAmpersandToken,
					SyntaxKind.BarBarToken,
				].includes((toks[i + 1] as Tok | undefined)?.kind as SyntaxKind)) ||
			// && C / || C
			([SyntaxKind.AmpersandAmpersandToken, SyntaxKind.BarBarToken].includes(
				t.kind,
			) &&
				isCarrierAt(i + 1));
		if (truthyHere) {
			push(
				i,
				"ceiling",
				"branches on the truthiness of an effects result; a Proxy cannot trap truthiness -- read ctx.dryRun, or forward the value instead",
			);
			continue;
		}

		// --- Proxy ceiling 2: identity comparison ---
		const identityKinds = [
			SyntaxKind.EqualsEqualsEqualsToken,
			SyntaxKind.ExclamationEqualsEqualsToken,
			SyntaxKind.EqualsEqualsToken,
			SyntaxKind.ExclamationEqualsToken,
		];
		if (
			identityKinds.includes(t.kind) &&
			(isCarrierAt(i - 1) || isCarrierAt(i + 1))
		) {
			push(
				i,
				"ceiling",
				"compares an effects result with ===; a Proxy cannot trap === -- forward the value instead",
			);
		}
	}
	return findings;
}

/**
 * Finds direct effect calls and untrappable carrier uses in sources under
 * `root`. Returns findings in file-then-line order.
 */
export function scanEffectsBypasses(root: string): BypassFinding[] {
	const findings: BypassFinding[] = [];
	let isDir = false;
	try {
		isDir = statSync(root).isDirectory();
	} catch {
		return findings;
	}
	if (!isDir) {
		return findings;
	}
	for (const path of collectSourceFiles(root)) {
		let text: string;
		try {
			text = readFileSync(path, "utf8");
		} catch {
			// A file the analyser cannot read is not evidence of a bypass.
			continue;
		}
		let toks: Tok[];
		try {
			toks = tokenize(text);
		} catch {
			continue;
		}
		findings.push(...scanTokens(toks, text, relative(root, path)));
	}
	return findings;
}

/**
 * The built-in provider. Registered whenever the check system turns on, so a
 * consumer that adopts checks at all gets the lint without a TOML declaration.
 */
export function effectsBypassProvider(_app: AppImpl): () => CheckSpec[] {
	// Named so tests can identify and drop the framework's own provider
	// (tests/helpers.ts dropBuiltinCheckProviders), mirroring Python's
	// conftest helper which filters on the provider function's __name__.
	return function effectsBypassCheckProvider(): CheckSpec[] {
		return [
			errorCheckSpec({
				name: "effects-bypass",
				tags: ["effects", "quality"],
				fast: true,
				pure: true,
				needsNetwork: false,
				dependsOn: [],
				impl: (ctx, reporter) => {
					const findings = scanEffectsBypasses(ctx.projectRoot);
					for (const f of findings) {
						reporter.error(`${f.file}:${f.line}: ${f.text}`);
					}
					if (findings.length === 0) {
						return reporter.passed("no direct effect calls bypass ctx.effects");
					}
					const direct = findings.filter((f) => f.kind === "call").length;
					const ceilings = findings.length - direct;
					if (ceilings === 0) {
						return reporter.found(
							`${direct} direct effect call(s) bypassing ctx.effects`,
						);
					}
					return reporter.found(
						`${direct} direct effect call(s) bypassing ctx.effects and ` +
							`${ceilings} untrappable carrier use(s)`,
					);
				},
			}),
		];
	};
}

File: go/strictcli/effects_bypass.go
package strictcli

// The effects-bypass lint (the `effects-bypass` check provider).
//
// Closed lists, matched on the called function/selector name: process starts,
// filesystem mutations, and network calls. The analyser is the stdlib go/ast +
// go/parser -- a regular dependency, no optional import and no soft
// degradation.

import (
	"fmt"
	"go/ast"
	"go/parser"
	"go/token"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// bypassProcess is the closed list of process-starting members. They are
// findings only when reached through a receiver (os/exec.Command, syscall.Exec),
// so an ordinary local Run() helper is not flagged by name alone.
var bypassProcess = map[string]bool{
	"Command": true, "CommandContext": true, "Start": true, "Output": true,
	"CombinedOutput": true, "StartProcess": true, "Exec": true,
	"ForkExec": true,
}

// bypassProcessReceivers are the receivers that make a bypassProcess member a
// finding.
var bypassProcessReceivers = map[string]bool{
	"exec": true, "syscall": true, "os": true, "cmd": true,
}

// bypassFilesystem is the closed list of filesystem-mutating members.
var bypassFilesystem = map[string]bool{
	"Create": true, "CreateTemp": true, "WriteFile": true, "Remove": true,
	"RemoveAll": true, "Mkdir": true, "MkdirAll": true, "MkdirTemp": true,
	"Rename": true, "Chmod": true, "Chown": true, "Truncate": true,
	"Symlink": true, "Link": true, "OpenFile": true,
}

// bypassNetwork is the closed list of network members, banned only when reached
// through one of the receivers below so an ordinary mapping.Get is not a
// finding.
var bypassNetwork = map[string]bool{
	"Get": true, "Post": true, "PostForm": true, "Head": true, "Do": true,
	"Dial": true, "DialTimeout": true, "NewRequest": true,
	"NewRequestWithContext": true,
}

var bypassNetworkReceivers = map[string]bool{
	"http": true, "net": true, "client": true, "Client": true,
	"DefaultClient": true, "transport": true, "ws": true,
}

// bypassSkipDirs are never walked.
var bypassSkipDirs = map[string]bool{
	".git": true, "vendor": true, "node_modules": true, "testdata": true,
	"dist": true, "build": true,
}

// bypassFinding is one direct effect call inside a function that opted into
// ctx.Effects().
type bypassFinding struct {
	file   string
	line   int
	fn     string
	target string
}

// effectsBypassProvider is the built-in check provider for the effects-bypass
// lint. It is registered whenever the check system turns on, so a consumer that
// adopts checks at all gets the lint without a TOML declaration. It fails on any
// direct process, filesystem-mutation or network call made from a function that
// opted into ctx.Effects().
func (a *App) effectsBypassProvider() []CheckSpec {
	impl := func(ctx CheckContext, reporter *ErrorReporter) CheckOutcome {
		findings := scanEffectsBypasses(ctx.ProjectRoot())
		for _, f := range findings {
			reporter.Error(fmt.Sprintf("%s:%d: %s calls %s directly; route it through ctx.Effects()",
				f.file, f.line, f.fn, f.target))
		}
		if len(findings) > 0 {
			return reporter.Found(fmt.Sprintf("%d direct effect call(s) bypassing ctx.Effects()", len(findings)))
		}
		return reporter.Passed("no direct effect calls bypass ctx.Effects()")
	}

	return []CheckSpec{
		NewErrorCheckSpec(CheckSpecMeta{
			Name:         "effects-bypass",
			Tags:         []string{"effects", "quality"},
			Severity:     "error",
			Fast:         true,
			Pure:         true,
			NeedsNetwork: false,
			DependsOn:    []string{},
		}, impl),
	}
}

// bypassFunc is one analysable function-like node, with the file it came from.
type bypassFunc struct {
	name    string
	body    *ast.BlockStmt
	rel     string
	pkgDir  string
	aliases map[string]bool
}

// scanEffectsBypasses finds direct effect calls REACHABLE FROM A REGISTERED
// COMMAND HANDLER. Results are in file then line order.
//
// §11's scope is reachability, not "a function whose own body mentions
// Effects()" -- a handler that never touches the handle, and a bypass one
// helper-call away, are both trivial escapes from the narrower reading, and this
// lint is the sole stated mitigation for the accepted no-sandbox ceiling.
//
// Roots are handler-shaped functions (first parameter *Context /
// *strictcli.Context, which is exactly the command-handler and
// passthrough-handler signature) plus, as before, any function that reaches for
// Effects() itself. From each root the closure follows DIRECT calls to
// package-level functions, transitively, WITHIN ONE PACKAGE (one directory) --
// the most go/ast can resolve without a type checker, and the boundary at which
// a bare name stops being unambiguous.
func scanEffectsBypasses(root string) []bypassFinding {
	var findings []bypassFinding
	info, err := os.Stat(root)
	if err != nil || !info.IsDir() {
		return findings
	}
	var files []string
	_ = filepath.Walk(root, func(path string, fi os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if fi.IsDir() {
			base := filepath.Base(path)
			if path != root && (bypassSkipDirs[base] || strings.HasPrefix(base, ".")) {
				return filepath.SkipDir
			}
			return nil
		}
		if strings.HasSuffix(path, ".go") {
			files = append(files, path)
		}
		return nil
	})
	sort.Strings(files)

	fset := token.NewFileSet()
	// Pass 1: parse every file, index package-level funcs per directory, and
	// collect every function-like node with its roots-ness precomputed.
	var all []*bypassFunc
	roots := map[*bypassFunc]bool{}
	// pkgDir -> func name -> the package-level function
	pkgFuncs := map[string]map[string]*bypassFunc{}
	for _, path := range files {
		tree, err := parser.ParseFile(fset, path, nil, 0)
		if err != nil {
			// A file the analyser cannot read is not evidence of a bypass.
			continue
		}
		rel, relErr := filepath.Rel(root, path)
		if relErr != nil {
			rel = path
		}
		pkgDir := filepath.Dir(path)
		fileAliases := effectsHandleAliases(tree)
		for _, decl := range tree.Decls {
			fd, ok := decl.(*ast.FuncDecl)
			if !ok || fd.Body == nil || fd.Recv != nil {
				continue
			}
			if _, seen := pkgFuncs[pkgDir]; !seen {
				pkgFuncs[pkgDir] = map[string]*bypassFunc{}
			}
			if _, dup := pkgFuncs[pkgDir][fd.Name.Name]; !dup {
				pkgFuncs[pkgDir][fd.Name.Name] = nil // reserved; filled below
			}
		}
		ast.Inspect(tree, func(n ast.Node) bool {
			name, body := functionNameAndBody(n)
			if body == nil {
				return true
			}
			bf := &bypassFunc{
				name:    name,
				body:    body,
				rel:     rel,
				pkgDir:  pkgDir,
				aliases: fileAliases,
			}
			all = append(all, bf)
			if isHandlerShaped(n) || functionOptsIntoEffects(body) {
				roots[bf] = true
			}
			if fd, ok := n.(*ast.FuncDecl); ok && fd.Recv == nil {
				if m := pkgFuncs[pkgDir]; m != nil {
					if existing, seen := m[fd.Name.Name]; seen && existing == nil {
						m[fd.Name.Name] = bf
					}
				}
			}
			return true
		})
	}

	// Pass 2: transitive closure over direct calls to package-level functions.
	reachable := map[*bypassFunc]bool{}
	var queue []*bypassFunc
	for _, bf := range all {
		if roots[bf] {
			queue = append(queue, bf)
		}
	}
	for len(queue) > 0 {
		bf := queue[len(queue)-1]
		queue = queue[:len(queue)-1]
		if reachable[bf] {
			continue
		}
		reachable[bf] = true
		for _, callee := range directCallNames(bf.body) {
			if target := pkgFuncs[bf.pkgDir][callee]; target != nil && !reachable[target] {
				queue = append(queue, target)
			}
		}
	}
	if len(reachable) == 0 {
		return findings
	}

	// Pass 3: report banned calls, once per call site, at the innermost
	// reachable enclosing function. Nesting means a call can sit inside several
	// analysable bodies; the innermost one names the finding.
	seen := map[token.Pos]bool{}
	for _, bf := range all {
		if !reachable[bf] {
			continue
		}
		ast.Inspect(bf.body, func(inner ast.Node) bool {
			if inner != ast.Node(bf.body) {
				if _, nested := functionNameAndBody(inner); nested != nil {
					// A nested function is its own bypassFunc; it reports its
					// own calls (and inherits reachability through `all`).
					if isNestedAnalysable(inner, all, reachable) {
						return false
					}
				}
			}
			call, ok := inner.(*ast.CallExpr)
			if !ok {
				return true
			}
			if seen[call.Pos()] {
				return true
			}
			if reachesEffectsHandle(call.Fun, bf.aliases) {
				return true
			}
			target, receiver := callTargetName(call.Fun)
			if target == "" {
				return true
			}
			leaf := target
			if i := strings.LastIndex(target, "."); i >= 0 {
				leaf = target[i+1:]
			}
			banned := (bypassProcess[leaf] && bypassProcessReceivers[receiver]) ||
				(bypassFilesystem[leaf] && receiver != "") ||
				(bypassNetwork[leaf] && bypassNetworkReceivers[receiver])
			if banned {
				seen[call.Pos()] = true
				findings = append(findings, bypassFinding{
					file:   bf.rel,
					line:   fset.Position(call.Pos()).Line,
					fn:     bf.name,
					target: target,
				})
			}
			return true
		})
	}
	sort.SliceStable(findings, func(i, j int) bool {
		if findings[i].file != findings[j].file {
			return findings[i].file < findings[j].file
		}
		return findings[i].line < findings[j].line
	})
	return findings
}

// isNestedAnalysable reports whether a nested function node is itself one of the
// reachable bypassFuncs, so the enclosing walk can stop and let it report.
func isNestedAnalysable(n ast.Node, all []*bypassFunc, reachable map[*bypassFunc]bool) bool {
	_, body := functionNameAndBody(n)
	for _, bf := range all {
		if bf.body == body {
			return reachable[bf]
		}
	}
	return false
}

// isHandlerShaped reports whether a function's FIRST parameter is *Context (or
// *strictcli.Context) -- exactly the command-handler and passthrough-handler
// signatures, and nothing else in the surface.
func isHandlerShaped(n ast.Node) bool {
	var params *ast.FieldList
	switch fn := n.(type) {
	case *ast.FuncDecl:
		params = fn.Type.Params
	case *ast.FuncLit:
		params = fn.Type.Params
	default:
		return false
	}
	if params == nil || len(params.List) == 0 {
		return false
	}
	star, ok := params.List[0].Type.(*ast.StarExpr)
	if !ok {
		return false
	}
	switch t := star.X.(type) {
	case *ast.Ident:
		return t.Name == "Context"
	case *ast.SelectorExpr:
		return t.Sel.Name == "Context"
	}
	return false
}

// directCallNames returns the bare `name(...)` callees inside a body.
func directCallNames(body *ast.BlockStmt) []string {
	var names []string
	ast.Inspect(body, func(n ast.Node) bool {
		if call, ok := n.(*ast.CallExpr); ok {
			if ident, ok := call.Fun.(*ast.Ident); ok {
				names = append(names, ident.Name)
			}
		}
		return true
	})
	return names
}

// effectsHandleAliases returns the names bound to the effects handle anywhere in
// a file: locals assigned from ctx.Effects() (`e := ctx.Effects()`) and
// parameters declared *Effects (a helper that takes the handle). Both are
// ordinary ways to write handler code and must not read as bypasses. File scope,
// not function scope, so a closure that uses its enclosing function's handle is
// covered too.
func effectsHandleAliases(tree *ast.File) map[string]bool {
	aliases := map[string]bool{}
	record := func(lhs, rhs []ast.Expr) {
		for i, r := range rhs {
			if i >= len(lhs) || !reachesEffectsHandle(r, nil) {
				continue
			}
			if ident, ok := lhs[i].(*ast.Ident); ok {
				aliases[ident.Name] = true
			}
		}
	}
	ast.Inspect(tree, func(n ast.Node) bool {
		switch s := n.(type) {
		case *ast.AssignStmt:
			record(s.Lhs, s.Rhs)
		case *ast.ValueSpec:
			lhs := make([]ast.Expr, len(s.Names))
			for i, name := range s.Names {
				lhs[i] = name
			}
			record(lhs, s.Values)
		case *ast.Field:
			if star, ok := s.Type.(*ast.StarExpr); ok && isEffectsTypeName(star.X) {
				for _, name := range s.Names {
					aliases[name.Name] = true
				}
			}
		}
		return true
	})
	return aliases
}

// isEffectsTypeName reports whether a type expression names the Effects handle.
func isEffectsTypeName(expr ast.Expr) bool {
	switch t := expr.(type) {
	case *ast.Ident:
		return t.Name == "Effects"
	case *ast.SelectorExpr:
		return t.Sel.Name == "Effects"
	}
	return false
}

// functionNameAndBody returns a function-like node's display name and body.
func functionNameAndBody(n ast.Node) (string, *ast.BlockStmt) {
	switch fn := n.(type) {
	case *ast.FuncDecl:
		return fn.Name.Name, fn.Body
	case *ast.FuncLit:
		return "func literal", fn.Body
	}
	return "", nil
}

// functionOptsIntoEffects reports whether a function body reaches for an
// Effects handle at all. One of the two root conditions: a function that uses
// the effects handle must route ALL of its effects through it, or the preview
// it promises is a lie.
func functionOptsIntoEffects(body *ast.BlockStmt) bool {
	found := false
	ast.Inspect(body, func(n ast.Node) bool {
		if sel, ok := n.(*ast.SelectorExpr); ok && sel.Sel.Name == "Effects" {
			found = true
			return false
		}
		return true
	})
	return found
}

// reachesEffectsHandle reports whether a callee's receiver chain goes through
// .Effects(). `aliases` are local names bound to the handle (`e :=
// ctx.Effects()`), which is an ordinary way to write a handler.
func reachesEffectsHandle(expr ast.Expr, aliases map[string]bool) bool {
	for {
		switch e := expr.(type) {
		case *ast.SelectorExpr:
			if e.Sel.Name == "Effects" {
				return true
			}
			expr = e.X
		case *ast.CallExpr:
			expr = e.Fun
		case *ast.Ident:
			return aliases[e.Name]
		default:
			return false
		}
	}
}

// callTargetName returns (dottedTarget, receiver) for a call's callee.
func callTargetName(expr ast.Expr) (string, string) {
	switch e := expr.(type) {
	case *ast.Ident:
		return e.Name, ""
	case *ast.SelectorExpr:
		parts := []string{e.Sel.Name}
		cur := e.X
		for {
			sel, ok := cur.(*ast.SelectorExpr)
			if !ok {
				break
			}
			parts = append(parts, sel.Sel.Name)
			cur = sel.X
		}
		if ident, ok := cur.(*ast.Ident); ok {
			parts = append(parts, ident.Name)
		}
		for i, j := 0, len(parts)-1; i < j; i, j = i+1, j-1 {
			parts[i], parts[j] = parts[j], parts[i]
		}
		receiver := ""
		if len(parts) >= 2 {
			receiver = parts[len(parts)-2]
		}
		return strings.Join(parts, "."), receiver
	}
	return "", ""
}
